src/openrct2/paint/Paint.cpp

```diff
@@ -360,125 +346,159 @@ static PaintStruct* PaintArrangeStructsHelperRotation(PaintStruct* ps_next, uint
             ps->SortFlags = flag | PaintSortFlags::PendingVisit;
         }
     } while (ps->quadrant_index <= quadrantIndex + 1);
-    ps = ps_temp;
+}
 
-    // Iterate all nodes in the current list and re-order them based on
-    // the current rotation and their bounding box.
+static std::pair<PaintStruct*, PaintStruct*> PaintStructsGetNextPending(PaintStruct* ps)
+{
+    PaintStruct* ps_next;
     while (true)
     {
-        // Get the first pending node in the quadrant list
-        while (true)
+        ps_next = ps->next_quadrant_ps;
+        if (ps_next == nullptr)
         {
-            ps_next = ps->next_quadrant_ps;
-            if (ps_next == nullptr)
-            {
-                // End of the current list.
-                return psQuadrantEntry;
-            }
-            if (ps_next->SortFlags & PaintSortFlags::OutsideQuadrant)
-            {
-                // Reached point outside of specified quadrant.
-                return psQuadrantEntry;
-            }
-            if (ps_next->SortFlags & PaintSortFlags::PendingVisit)
-            {
-                // Found node to check on.
-                break;
-            }
-            ps = ps_next;
+            // End of the current list.
+            return { nullptr, nullptr };
+        }
+        if (ps_next->SortFlags & PaintSortFlags::OutsideQuadrant)
+        {
+            // Reached point outside of specified quadrant.
+            return { nullptr, nullptr };
+        }
+        if (ps_next->SortFlags & PaintSortFlags::PendingVisit)
+        {
+            // Found node to check on.
+            break;
+        }
+        ps = ps_next;
+    }
+    return { ps, ps_next };
+}
+
+template<uint8_t TRotation> static void PaintStructsSortQuadrant(PaintStruct* parent, PaintStruct* child)
+{
+    // Mark visited.
+    child->SortFlags &= ~PaintSortFlags::PendingVisit;
+
+    // Compare all the children below the first child and move them up in the list if they intersect.
+    const PaintStructBoundBox& initialBBox = child->bounds;
+    for (;;)
+    {
+        auto* ps = child;
+        child = child->next_quadrant_ps;
+
+        if (child == nullptr || child->SortFlags & PaintSortFlags::OutsideQuadrant)
+        {
+            break;
         }
 
-        // Mark visited.
-        ps_next->SortFlags &= ~PaintSortFlags::PendingVisit;
-        ps_temp = ps;
+        if (!(child->SortFlags & PaintSortFlags::Neighbour))
+        {
+            continue;
+        }
```

Comment:
The code prior was following:
```cpp
        if (child->SortFlags & PaintSortFlags::OutsideQuadrant)
            break;
        if (!(child->SortFlags & PaintSortFlags::Neighbour))
            continue;
```
If we change the order to:
```cpp
        if (!(child->SortFlags & PaintSortFlags::Neighbour))
            continue;
        if (child->SortFlags & PaintSortFlags::OutsideQuadrant)
            break;
```
We will have a 20 FPS performance regression, some magical msvc compiler bug. I was pulling my hair out for a while trying to figure out what caused it, the idea was to order the conditions based on likely hood but that seems to screw with the optimizations somehow, so we should never change this despite the likelyhood hitting being outside the quadrant being nearly zero.